### PR TITLE
Feature/#26 forum posts list

### PR DESCRIPTION
**Pull request by vihopec**
Added:
- posts list
- see more logic (button)
- logic to sort posts (latest, popular, older)

desktop | mobile | spinner
:-------------------------:|:-------------------------:|:-------------------------
Screenshot 2021-11-11 at 09 50 33 | mobile version 1 | spinner on loading posts
desktop version 2| Screenshot 2021-11-11 at 10 02 24

### CODE REVIEW

File: src/pages/posts/SearchFilterParams/index.tsx [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+  {
+    label: 'Older',
+    value: 'oldest',
+  },
+]
+
+export const SearchBar = () => {
+  const dispatch = useDispatch()
+  const handleSortTypeClick = (event: React.ChangeEvent<HTMLInputElement>) => {
+    // eslint-disable-next-line @typescript-eslint/ban-ts-comment
```
Comment (vihopec) [inline src/pages/posts/SearchFilterParams/index.tsx:27]: I will fix it

Comment (Chubatik) [reply src/pages/posts/SearchFilterParams/index.tsx:27]: and don't forget about types)

File: src/store/postsSlice.tsx [hunk 2]
```diff
@@ -0,0 +1,73 @@
+import { createSlice, PayloadAction } from '@reduxjs/toolkit'
+
+import { PostsService } from '../API/PostsService'
+import { IPostsParams, ISinglePostResult, ISortParams } from '../models/SinglePostResult'
+import { POSTS_PER_PAGE } from '../utils/constants'
+
+import { RootState, AppDispatch } from './store'
+
+interface IPostsState {
```
Comment (YanaGushlevskaya) [inline src/store/postsSlice.tsx:9]: Lets add also an error to the state and action to handle it.

File: src/store/postsSlice.tsx [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+    },
+    setIsLoading: (state, action: PayloadAction<boolean>) => {
+      state.isLoading = action.payload
+    },
+  },
+})
+
+export const { addPosts, setPagesAmount, setSortType, setCurrentPage, setIsLoading } = postsSlice.actions
+
+export const selectPosts = (state: RootState): ISinglePostResult[] => state.posts.data
```
Comment (YanaGushlevskaya) [inline src/store/postsSlice.tsx:51]: How about adding selectors and loadPostsList/loadPages to different files? (postsSelectors, postEffects)? As a proposal.

File: src/store/postsSlice.tsx (hunk 2 above)
Comment (vihopec) [reply src/store/postsSlice.tsx:9]: How should we handle the error?

Comment (YanaGushlevskaya) [reply src/store/postsSlice.tsx:9]: Add action 'setError', error to state, then add .catch to loadPostsList and loadPageNumbers to dispatch 'setError'.
On front we can check if an error exists, then show something (for example toaster messages with some default message).

File: src/pages/posts/SearchFilterParams/index.tsx (hunk 1 above)
Comment (vihopec) [reply src/pages/posts/SearchFilterParams/index.tsx]: fixed

File: src/components/post/Post.module.scss [hunk 4]
```diff
@@ -2,12 +2,12 @@
...
 .container {
-  max-width: 280px;
-  margin: 10px 0;
+  width: 100%;
+  margin-bottom: 20px;
   box-shadow: 0px 8px 16px 0px rgba(34, 60, 80, 0.2);
   @include medium {
     min-width: 450px;
-    margin: 20px 0;
+    margin-bottom: 20px;
```
[the remark this answers is not in the document]
Comment (vihopec) [reply src/components/post/Post.module.scss]: updated

File: src/utils/constants.ts [hunk 5]
```diff
@@ -7,3 +7,5 @@ export const ROUTES = {
   SINGLE_POST: '/posts/:id',
   PROFILE: '/profile/:username',
 }
+
+export const POSTS_PER_PAGE = 2
```
Comment (Chubatik) [inline src/utils/constants.ts:11]: Please, resolve conflict in this file

File: src/store/postsSlice.tsx (hunk 2 above)
Comment (vihopec) [reply src/store/postsSlice.tsx:10]: done

File: src/store/postsSlice.tsx (hunk 3 above)
Comment (vihopec) [reply src/store/postsSlice.tsx]: created separate files for selectors

File: src/pages/posts/Posts.tsx [hunk 6]
```diff
@@ -1,28 +1,89 @@
...
+            <>
+              {currentPage < pagesAmount && (
+                <div className={classes.forum__posts__seeMoreButton}>
+                  <Button primary onClick={handleLoadingPosts} className={classes.forum__loadingButton}>
+                    See more
+                  </Button>
+                </div>
+              )}
+            </>
+          )}
```
Comment (irenlian) [inline src/pages/posts/Posts.tsx:79]:
```suggestion
          {isLoading && (
            <div className={classes.forum__posts__loadingSpinner}>
              <Spinner />
            </div>
          )}
          {!isLoading && currentPage < pagesAmount && (
                <div className={classes.forum__posts__seeMoreButton}>
                  <Button primary onClick={handleLoadingPosts} className={classes.forum__loadingButton}>
                    See more
                  </Button>
                </div>
          )}
```

File: src/store/selectors/posts.ts [hunk 7]
```diff
@@ -0,0 +1,12 @@
+import { ISinglePostResult, ISortParams } from '../../models/SinglePostResult'
+import { RootState } from '../store'
+
+export const selectPosts = (state: RootState): ISinglePostResult[] => state.posts.data
```
Comment (irenlian) [inline src/store/selectors/posts.ts:4]:
```suggestion
export const selectPosts = ({ posts }: RootState): ISinglePostResult[] => posts.data
```

File: src/store/store.ts [hunk 8]
```diff
@@ -1,11 +1,14 @@
...
 
 export const store = configureStore({
   reducer: {
     user: userSlice.reducer,
+    posts: postsSlice.reducer,
   },
 })
 
 export type RootState = ReturnType<typeof store.getState>
+export type AppDispatch = typeof store.dispatch
```
Comment (irenlian) [inline src/store/store.ts:14]: I assume there should already exist a type for the library's dispatch attribute.🤔

**Review by irenlian — APPROVED** [review]
Looks good. Just minor comments:

**Review by YanaGushlevskaya — APPROVED** [review]

File: src/store/store.ts (hunk 8 above)
Comment (vihopec) [reply src/store/store.ts:14]: I copied this approach from the documentation

https://react-redux.js.org/using-react-redux/usage-with-typescript#typing-the-usedispatch-hook

File: src/store/selectors/posts.ts (hunk 7 above)
Comment (vihopec) [reply src/store/selectors/posts.ts]: updated

File: src/pages/posts/Posts.tsx (hunk 6 above)
Comment (vihopec) [reply src/pages/posts/Posts.tsx]: updated

File: src/store/postsSlice.tsx [hunk 9]
```diff
@@ -0,0 +1,73 @@
...
+
+export const selectCurrentPage = (state: RootState): number => state.posts.currentPage
+
+export const selectIsLoading = (state: RootState): boolean => state.posts.isLoading
+
+export const loadPostsList = (params: IPostsParams) => (dispatch: AppDispatch) => {
+  dispatch(setIsLoading(true))
+  return PostsService.getPosts(params).then((res) => {
+    dispatch(addPosts(res.data))
+    dispatch(setIsLoading(false))
```
[the remark this answers is not in the document]
Comment (vihopec) [reply src/store/postsSlice.tsx]: done

**Pull request merged**